monai/inferers/splitter.py

```diff
@@ -56,144 +57,210 @@ def __call__(self, inputs: Any) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
+            self.reader = WSIReader(backend=reader, **self.reader_kwargs)
+        elif isclass(reader) and issubclass(reader, BaseWSIReader):
+            self.reader = reader(**self.reader_kwargs)
+        elif isinstance(reader, BaseWSIReader):
+            self.reader = reader
+        elif reader is not None:
+            raise ValueError(f"Unsupported reader type: {reader}.")
 
     def _calculate_pad_size(self, spatial_shape, spatial_ndim, patch_size, offset, overlap):
-        if not self.pad_mode:
+        if not self.pad_kwargs["mode"]:
             return [], False
         # initialize with zero
         pad_size = [0] * 2 * spatial_ndim
         # set the starting pad size only if the offset is negative
         pad_size[1::2] = (-min(off, 0) for off in offset)
         # set the ending pad size only if it is not divisible by the patch size
         pad_size[::2] = (
-            0 if ps == 0 else (off - ins + ps) % round(ps * (1.0 - ov))
-            for ins, off, ps, ov in zip(spatial_shape, offset, patch_size, overlap)
+            0 if ps == 0 else (off - sh + ps) % round(ps - (ps * ov if isinstance(ov, float) else ov))
+            for sh, off, ps, ov in zip(spatial_shape, offset, patch_size, overlap)
         )
         return pad_size, any(pad_size[1::2])
 
-    def __call__(self, inputs: torch.Tensor) -> Iterable[tuple[torch.Tensor, Sequence[int]]]:
+    def _get_valid_shape_parameters(
+        self, spatial_shape: Sequence[int]
+    ) -> tuple[tuple[int, ...], tuple[float, ...] | tuple[int, ...], tuple[int, ...]]:
+        spatial_ndim = len(spatial_shape)
+        # patch_size
+        patch_size = ensure_tuple_rep(self.patch_size, spatial_ndim)
+        # overlap
+        overlap = ensure_tuple_rep(self.overlap, spatial_ndim)
+        overlap = tuple(o if p else type(overlap[0])(0) for o, p in zip(overlap, patch_size))
+        # offset
+        offset = ensure_tuple_rep(self.offset, spatial_ndim)
+        for off, ps, sh in zip(offset, patch_size, spatial_shape):
+            if off < -ps:
+                raise ValueError(f"Negative `offset` ({off}) cannot be larger than `patch_size` ({ps}) in magnitude.")
+            if off >= sh:
+                raise ValueError(f"`offset` ({off}) cannot be larger than inputs size ({sh}).")
+        return patch_size, overlap, offset
+
+    def _get_patch_tensor(
+        self, inputs: torch.Tensor, location: tuple[int, ...], patch_size: tuple[int, ...]
+    ) -> torch.Tensor:
+        slices = (slice(None),) * self.non_spatial_ndim + tuple(
+            slice(loc, loc + ps) for loc, ps in zip(location, patch_size)
+        )
+        patch = inputs[slices]
+        # send the patch to target device
+        patch.to(self.device)
+        return patch
+
+    def _get_patch_wsi(self, inputs: Any, location: tuple[int, int], patch_size: tuple[int, int]) -> torch.Tensor:
+        patch, _ = self.reader.get_data(wsi=inputs, location=location, size=patch_size)
```

Comment:
I see you need to pass `inputs`, `location`, `size` args to `WSIReader.get_data()`, but other readers don't have these args: https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L126
How do you handle it within SlidingWindowSlitter? Or maybe I misunderstand your idea.

Thanks.

Reply:
It is not implement in this PR but we can check if it is not wsi and call as following:
```python
if isinstance(self.reader, ImageReader) and not isinstance(self.reader, BaseWSIReader):
    img_obj = self.reader.read(inputs)
    processed_input, _ = reader.get_data(img_obj)
```
Then it will use `self._get_patch_tensor` instead.
https://github.com/Project-MONAI/MONAI/blob/c032036e16c196cf912827ca44513d24a337d35b/monai/inferers/splitter.py#L209

Please keep in mind that here we have two pathways, one for WSI and one for anythong else. WSI pathway is specific because the image does not fit into memory for the rest of image readers, the image fits into memory so it should go through the tensor path way after it is loaded.

Reply:
Thanks for your explanation.
@ericspod @wyli as WSI and the regular sliding window are 2 quite different pathways, I still feel it's better to extend the WSI feature in a subclass, what do you guys think?

Thanks.

Reply:
Hi @drbeh ,

I am open to a discussion, and hope we can get some input from @ericspod @wyli, and others.

Thanks.

Reply:
Thanks @Nic-Ma, with the changes that I made, now subclassing would make sense. 
We have several features in the backlog so it's best if we can move fast from here but thanks again for your constrictive comments.